### PR TITLE
Register candidates and invite applicants

### PR DESCRIPTION
**Pull request by henricook**
Still contains many "test" methods for use with the ManualClient, all will be removed before Prod.

Tested workflows involving only POSTs are:

- Create Candidate
- Invite Candidate

(See ApplicationController)

### CODE REVIEW

File: app/uk/gov/hmrc/fsetlaunchpadgateway/config/frontendWiring.scala [hunk 1]
```diff
@@ -10,10 +11,13 @@ object FrontendAuditConnector extends Auditing with AppName {
   override lazy val auditingConfig = LoadAuditingConfig(s"auditing")
 }
 
-object WSHttp extends WSGet with WSPut with WSPost with WSDelete with AppName with RunMode {
+object WSHttp extends WSHttp {
   override val hooks = NoneRequired
 }
 
+trait WSHttp extends WSGet with WSPutWithForms with WSPost with WSDelete with AppName with RunMode {
```
Comment (henricook) [inline app/uk/gov/hmrc/fsetlaunchpadgateway/config/frontendWiring.scala]: Make this more easily mockable

File: app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/ApplicationController.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        )
+      ).map(seamlessLoginInviteResponse => Ok(Json.toJson(seamlessLoginInviteResponse))).recover(recoverFromBadCall)
+    }
+  }
+
+  private def recoverFromBadCall: PartialFunction[Throwable, Result] = {
+    case e: Throwable =>
+      Logger.warn(s"Error communicating with launchpad: ${e.getMessage}. Stacktrace: ${e.getStackTrace}")
+      InternalServerError("Error communicating with Launchpad")
+    case _ => InternalServerError
```
Comment (ollyw) [inline app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/ApplicationController.scala]: I don't believe this case isn't used as the Partial is a [Throwable, Result] and you are matching Throwable

File: app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/CallbackController.scala [hunk 3]
```diff
@@ -23,3 +23,10 @@ trait CallbackController extends FrontendController {
     }
   }
 }
+
+// scalastyle:off
+/*
+
+  {"app":"fset-launchpad-gateway","hostname":"public-app-13-fset-launchpad-gateway","timestamp":"2016-08-22 15:23:44.331+00:00","message":"Received callback => AnyContentAsJson({\"candidate_id\":\"cnd_4bf18c2e67fbdfe122614c894403ad1a\",\"custom_candidate_id\":\"CUSTOM_CSR_ID_TO_KNOW_WHO_IS_WHO\",\"interview_id\":13755,\"custom_interview_id\":null,\"custom_invite_id\":\"CSR_CUSTOM_INVITE_REFERENCE\",\"status\":\"final\"})\n\nWith static key = foobar123123123123123123123123123","logger":"application","thread":"play-akka.actor.default-dispatcher-557","level":"INFO","application.home":"/app/fset-launchpad-gateway-0.3.0"}
```
Comment (ollyw) [inline app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/CallbackController.scala]: If this is still needed, a comment may help the befuddled

File: project/FrontendBuild.scala [hunk 4]
```diff
@@ -11,22 +11,23 @@ object FrontendBuild extends Build with MicroService {
...
-  val jsonLogger    = "2.1.1"
-  val passcode      = "3.2.0"
-  val scalatest     = "2.2.2"
-  val pegdown       = "1.4.2"
-  val jsoup         = "1.7.3"
-  val wiremock      = "1.57"
-  val hmrctest      = "1.4.0"
-  val scalatestplus = "1.2.0"
-  val silhouette    = "2.0.2"
+  val microserviceBootstrapVersion  = "4.2.0"
```
Comment (ollyw) [inline project/FrontendBuild.scala]: I take it that we are not using this? This is a frontend, so shouldn't need the microservice too

File: test/uk/gov/hmrc/fsetlaunchpadgateway/connectors/launchpad/ClientSpec.scala [hunk 5]
```diff
@@ -1,120 +1,194 @@
...
+case class ClientTestException(message: String, stringsToRemove: List[String]) extends SanitizedClientException(message, stringsToRemove)
 
-  "Testing" should {
-    "Client test" ignore {
+case class ClientTestRequest(
+  firstName: String,
+  lastName: String,
+  email: String
+) extends ContainsSensitiveData {
+  override def getSensitiveStrings: List[String] = List(firstName, email)
```
Comment (ollyw) [inline test/uk/gov/hmrc/fsetlaunchpadgateway/connectors/launchpad/ClientSpec.scala]: Lastname is sensitive if first name is

Comment (henricook) [inline test/uk/gov/hmrc/fsetlaunchpadgateway/connectors/launchpad/ClientSpec.scala]: Yep, it wasn't included here purely so i could have one field not obfuscated for testing purposes

File: project/FrontendBuild.scala (hunk 4 above)
Comment (henricook) [inline project/FrontendBuild.scala]: We are, it gives us something to do with controllers if i recall correctly, we also have it in cubiks-gateway

File: test/uk/gov/hmrc/fsetlaunchpadgateway/connectors/launchpad/ClientSpec.scala (hunk 5 above)
Comment (ollyw) [inline test/uk/gov/hmrc/fsetlaunchpadgateway/connectors/launchpad/ClientSpec.scala]: You are right. Should have seen the file name! Wish IntelliJ had github review integration :-)

File: app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/CallbackController.scala (hunk 3 above)
Comment (henricook) [inline app/uk/gov/hmrc/fsetlaunchpadgateway/controllers/CallbackController.scala]: Added a description to it

File: project/FrontendBuild.scala (hunk 4 above)
Comment (henricook) [inline project/FrontendBuild.scala]: BaseController which gives us withJsonBody[T] is from the microservice-bootstrap package

**Comment by henricook** [comment]
@lukastymo comments addressed and the launchpad gateway PR has been updated to accomodate the new camelCased response objects

**Comment by henricook** [comment]
@ollyw Comments addressed

**Comment by ollyw** [comment]
@henricook looks good. It's going in...

**Pull request merged**